drain3/template_miner.py

```diff
@@ -170,24 +174,129 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
+
+        For most accurate results, it is recommended that
+        - Each `MaskingInstruction` has a unique `mask_with` value,
+        - No `MaskingInstruction` has a `mask_with` value of `*`,
+        - The regex-patterns of `MaskingInstruction` do not use unnamed back-references;
+          instead use back-references to named groups e.g. `(?P=some-name)`.
+
+        :param log_template: log template corresponding to the log message
+        :param log_message: log message to extract parameters from
+        :param exact_matching: whether to apply the correct masking-patterns to match parameters, or try to approximate;
+            disabling exact_matching may be faster but may lead to situations in which parameters
+            are wrongly identified.
+        :return: A ordered list of ExtractedParameter for the log message
+            or None if log_message does not correspond to log_template.
+        """
+
+        for delimiter in self.config.drain_extra_delimiters:
+            log_message = re.sub(delimiter, " ", log_message)
+
+        template_regex, param_group_name_to_mask_name = self._get_template_parameter_extraction_regex(
+            log_template, exact_matching)
+
+        # Parameters are represented by specific named groups inside template_regex.
+        parameter_match = re.match(template_regex, log_message)
+
+        # log template does not match template
+        if not parameter_match:
+            return None
+
+        # create list of extracted parameters
+        extracted_parameters = []
+        for group_name, parameter in parameter_match.groupdict().items():
+            if group_name in param_group_name_to_mask_name:
+                mask_name = param_group_name_to_mask_name[group_name]
+                extracted_parameter = ExtractedParameter(parameter, mask_name)
+                extracted_parameters.append(extracted_parameter)
+
+        return extracted_parameters
+
+    @cachedmethod(lambda self: self.parameter_extraction_cache)
+    def _get_template_parameter_extraction_regex(self, log_template: str, exact_matching: bool):
+        param_group_name_to_mask_name = dict()
+        param_name_counter = [0]
+
+        def get_next_param_name():
+            param_group_name = "p_" + str(param_name_counter[0])
+            param_name_counter[0] += 1
+            return param_group_name
+
+        # Create a named group with the respective patterns for the given mask-name.
+        def create_capture_regex(_mask_name):
+            allowed_patterns = []
+            if exact_matching and _mask_name in self.masker.mask_names:
+                for mi in self.masker.instructions_by_mask_name(_mask_name):
+                    # MaskingInstruction may already contain named groups.
+                    # We replace group names in those named groups,
+                    # to avoid conflicts due to duplicate names.
+                    mi_groups = mi.regex.groupindex.keys()
```

Comment:
So the idea of using an AbstractMaskingInstruction is nice but this line breaks the encapsulation because we assume that it had a regex field. I think we have to revert it unless you have another idea..

Reply:
actually we can check if its not regex instruction and then don't use exact matching, but I don't know if we want to spend effort on that.

Reply:
Yes, this whole method will only work as long as `RegexMaskingInstruction` is the only `AbstractMaskingInstruction`.
I'm not sure how other instructions could even look like, but if they are added, a solution would need to be found based on whatever those look like.
By not checking, the developer of a new instruction is forced to deal with this. 

Let me explain my motivation behind the changes, it's purely subjective:
Previously `RegexMasker` existed along with a `LogMasker` which could theoretically use different maskers, implying a masker could be developed that does not rely on regexes. 
But the `MaskingInstruction`-objects themselves actually do all the work of masking using a regex.
By reframing it this way, a single `LogMasker` is needed, that just collects multiple instructions.
`AbstractMaskingInstruction` then specifies a minimal interface that `LogMasker` can work with.
As it is not clear how an instruction different from `MaskingInstruction` could look like, the interface of `AbstractMaskingInstruction` likely would need to be changed to work with `extract_parameters` in case a new instruction type is ever developed.

I'd say if respecting encapsulation is more important, let's just scrap `AbstractMaskingInstruction`.

Reply:
Another option might be to say that masking-patterns need to be able to be detected via a regex, even if they do not use regexes internally.
This would result in a method like `as_regex`added to `AbstractMaskingInstructions`, which `RegexMaskingInstruction` would simply satisfy by returning its compiled regex-pattern.

Reply:
I added support for that following my initial idea